codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/shapevisitor/DafnyToSmithyShapeVisitor.java

```diff
@@ -0,0 +1,686 @@
+        if %s == nil {
+                return nil
+        }""".formatted(dataSource);
+    }
+    final String functionInit =
+      """
+          var union %s
+          %s
+      """.formatted(context.symbolProvider().toSymbol(shape), nilCheck);
+    final StringBuilder eachMemberInUnion = new StringBuilder();
+    for (final var member : shape.getAllMembers().values()) {
+      final Shape targetShape = context.model().expectShape(member.getTarget());
+      final String memberName = context.symbolProvider().toMemberName(member);
+      final String rawUnionDataSource =
+        "(" +
+        dataSource +
+        ".(" +
+        DafnyNameResolver.getDafnyType(
+          shape,
+          context.symbolProvider().toSymbol(shape)
+        ) +
+        "))";
+      // unwrap union type, assert it then convert it to its member type with Dtor_ (example: Dtor_BlobValue()). unionDataSource is not a wrapper object until now.
+      String unionDataSource =
+        rawUnionDataSource +
+        ".Dtor_" +
+        memberName.replace(shape.getId().getName().concat("Member"), "") +
+        "()";
+      final boolean isMemberShapePointable =
+        (GoPointableIndex.of(context.model()).isPointable(member)) &&
+        !targetShape.isStructureShape();
+      final String pointerForPointableShape = isMemberShapePointable ? "*" : "";
+      final String isMemberCheck =
+        "if ((%s).%s()) {".formatted(
+            rawUnionDataSource,
+            memberName.replace(shape.getId().getName().concat("Member"), "Is_")
+          );
+      String wrappedDataSource = "";
+      boolean requireAssertion = true;
+      if (!(targetShape.isStructureShape())) {
+        // All other shape except structure needs a Wrapper object but unionDataSource is not a Wrapper object.
+        wrappedDataSource =
+          "var dataSource = Wrappers.Companion_Option_.Create_Some_(%s)".formatted(
+              unionDataSource
+            );
+        unionDataSource = "dataSource.UnwrapOr(nil)";
+        requireAssertion = false;
+      }
+      eachMemberInUnion.append(
+        """
+        %s
+        %s
+        union = &%s.%s{
+                Value: %s(%s),
+            }
+        }
+        """.formatted(
+            isMemberCheck,
+            wrappedDataSource,
+            SmithyNameResolver.smithyTypesNamespace(shape),
+            memberName,
+            pointerForPointableShape,
+            ShapeVisitorHelper.toNativeShapeVisitorWriter(
+              member,
+              context,
+              unionDataSource,
+              requireAssertion,
+              writer,
+              isConfigShape,
+              isMemberShapePointable
+            )
+          )
+      );
+    }
+    return """
+        %s
+        %s
+        return union
+    """.formatted(functionInit, eachMemberInUnion);
+  }
+
+  @Override
+  public String timestampShape(final TimestampShape shape) {
+    // TODO: Figure out timestamp types when working on timestampShape
```

Comment:
I would prefer to raise an exception at codegen time if this is encountered -- safer to make codegen fail on unhandled shapes

Reply:
AFAIK throwing exception fails the codegen on DDBv2 (new model specifically for Go) shapes, hence the `nil` part.

Reply:
Is `DafnyToSmithyShapeVisitor` used for both LocalService and AWS SDK shapes?

(Python only implements TimestampShape for AWS SDKs. It raises an exception at codegen time for LocalServices.)

Reply:
Ahh I should be able to do that. I'll fix it up in a follow up PR since it's going to be again about ShapeVisitors.